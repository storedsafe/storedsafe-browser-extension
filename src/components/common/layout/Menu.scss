@import '../../../styles/variables.scss';

.menu {
  max-height: 0;
  width: 100%;
  z-index: 1;
  overflow: hidden;

  &.show {
    width: 100%;
    max-height: 100%;
    transition: $transition-duration;
  }

  .menu-box {
    position: relative;
    overflow: hidden;
    display: flex;
    flex-direction: column;
  }

  .menu-item {
    display: flex;
    outline: 0;
    border: 0;
    background-color: $color-primary-3;
    border-bottom: 1px solid $color-primary-3;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    color: $color-fg-alt;
    padding: $spacing;
    transition: $transition-duration;

    &:focus,
    &:hover {
      background-color: $color-accent-0;
    }

    > svg {
      fill: $color-fg-alt;
      width: $menu-size - $spacing * 2;
      height: $menu-size - $spacing * 2;
      transition: $transition-duration;
    }

    &:hover > :last-child {
      transform: scale(1.1);
    }
  }
}
