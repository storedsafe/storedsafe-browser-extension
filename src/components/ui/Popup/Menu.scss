@import '../../../variables';

.menu {
  $padding: 11.25px;

  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: center;
  background-color: $color-primary-0;
  width: 100%;
  height: 100%;

  .menu-item {
    display: flex;
    flex-direction: row;
    align-items: center;
    transition: $transition-duration;
    transition-property: background-color, color, flex-grow;
    padding: $padding;
    color: $color-fg-alt;
    cursor: pointer;
    width: 75px - $padding;
    height: 100%;
    overflow: hidden;

    &:hover {
      flex-grow: 1;
    }

    &:hover {
      background-color: $color-bg;
      color: $color-accent-0;
    }

    &.selected {
      background-color: $color-bg-alt;
      color: $color-accent-0;
    }

    &:hover .menu-item-icon > svg,
    &.selected .menu-item-icon > svg {
      fill: $color-accent-0;
    }

    .menu-item-icon {
      height: 30px;
      width: 30px;
      margin: $padding / 2;
      margin-right: $padding / 2 + $padding;
    }

    .menu-item-icon > * {
      height: 30px;
      width: 30px;
    }

    .menu-item-icon > svg {
      fill: $color-bg;
      transition: fill $transition-duration;
    }
  }
}
