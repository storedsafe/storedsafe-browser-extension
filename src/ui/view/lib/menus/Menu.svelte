<script lang="ts">
  import type { Page, PageNavigation } from "@/ui/view/Popup/pages";
  import Icon from "@/ui/view/lib/layout/Icon.svelte";

  interface Props {
    menuItems: PageNavigation[];
    selected: string | null;
    onNavigate?: (page: Page) => void;
  }

  let { menuItems, selected = null, onNavigate }: Props = $props();

  // Track focused and hovered state for accurate coloring
  let focused: string | null = $state(null);
  let hovered: string | null = $state(null);

  // Currently highlighted or selected object, depending on hover/focus state
  let menuItem = $derived(
    menuItems.find((item) => item.route === (hovered ?? focused ?? selected))
  );

  // Set properties of menu icons
  const iconProps = {
    isButton: true,
  };

  /**
   * Clear focused state reference.
   * */
  function clearFocused() {
    focused = null;
    hovered = null;
  }

  /**
   * Set focused state reference.
   * Focused has higher priority than hover,
   * so clear hovered state when focus changes.
   * @param name Name identifier of focused element.
   * */
  function setFocused(name: string) {
    focused = name;
    hovered = null;
  }

  /**
   * Clear hovered state reference.
   * @param name Name identifier of focused element.
   * */
  function clearHovered() {
    hovered = null;
  }

  /**
   * Set hovered state reference.
   * @param name Name identifier of focused element.
   * */
  function setHovered(name: string) {
    hovered = name;
  }
</script>

<!--
  @component
  Top menu of extension.
  Emits `navigate` event with the menu item name on navigation.
-->
<nav>
  <section
    class="icons"
    class:has-selected={!!menuItem}
    onmouseleave={() => clearHovered}
    role="button"
    tabindex="0"
  >
    {#each menuItems as item (item.route)}
      <button
        type="button"
        class="input-reset"
        class:selected={selected === item.route}
        onmouseenter={() => setHovered(item.route)}
        onfocus={() => setFocused(item.route)}
        onclick={() => onNavigate?.(item.route)}
        onblur={() => clearFocused()}
        aria-label={item.title}
      >
        <Icon {...{ ...iconProps, d: item.icon }} />
      </button>
    {/each}
  </section>
  {#if !!menuItem}
    <section
      class:selected={selected === menuItem.route}
      class:focused={focused === menuItem.route}
      class:hovered={hovered === menuItem.route}
      class="menu-title shadow"
      title={menuItem.title}
    >
      {menuItem.title}
    </section>
  {/if}
</nav>

<style>
  nav {
    width: 100%;
  }

  button {
    width: 100%;
    padding: calc(var(--spacing) / 2);
    background-color: var(--color-primary-dark);
    border-top: 2px solid transparent;
    transition: background-color 0.2s;
  }

  button:first-child {
    border-top-left-radius: 4px;
    border-bottom-left-radius: 4px;
  }

  button:last-child {
    border-top-right-radius: 4px;
    border-bottom-right-radius: 4px;
  }

  section.has-selected button:first-child,
  section.has-selected button:last-child {
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
  }

  section.has-selected button {
    border-bottom-right-radius: 0;
    border-bottom-left-radius: 0;
    transition: background-color 0.2s;
  }

  button:focus,
  button:hover {
    border-top: 2px solid #fff;
  }

  button:hover,
  .menu-title.hovered,
  button:focus,
  .menu-title.focused {
    background-color: var(--color-accent-light);
  }

  button.selected:not(:hover):not(:focus),
  .menu-title.selected:not(.hovered):not(.focused) {
    background-color: var(--color-accent);
  }

  button:active {
    background-color: #fff;
  }

  .icons {
    display: flex;
    justify-content: space-evenly;
  }

  .menu-title {
    color: #fff;
    text-align: center;
    font-weight: initial;
    border-bottom-left-radius: 4px;
    border-bottom-right-radius: 4px;
    padding: calc(var(--spacing) / 2);
    transition: background-color 0.2s;
    user-select: none;
  }
</style>
