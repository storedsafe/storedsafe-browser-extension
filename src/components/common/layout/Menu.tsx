import React from 'react'
import './Menu.scss'

export interface MenuItem {
  title: string
  icon: React.ReactNode
  onClick: () => void
}

interface MenuProps {
  show: boolean
  items: MenuItem[]
  onClick?: () => void
}

export const Menu: React.FunctionComponent<MenuProps> = ({
  show,
  items,
  onClick: closeMenu
}: MenuProps) => {
  const onItemClick = (onClick: () => void) => {
    return (): void => {
      onClick()
      closeMenu?.()
    }
  }

  return (
    <section className={`menu${show ? ' show' : ''}`}>
      <div className='menu-backdrop'>
        <div className='menu-box'>
          {items.map(({ title, icon, onClick }) => (
            <article
              key={title}
              className='menu-item'
              onClick={onItemClick(onClick)}
            >
              <p>{title}</p>
              {icon}
            </article>
          ))}
        </div>
      </div>
    </section>
  )
}
