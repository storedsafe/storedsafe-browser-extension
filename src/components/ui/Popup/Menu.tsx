import React from 'react';
import './Menu.scss';

interface MenuItemProps {
  title: string;
  icon: React.ReactNode;
  onClick: () => void;
  selected?: boolean;
}

const MenuItem: React.FunctionComponent<MenuItemProps> = ({
  title,
  icon,
  onClick,
  selected
}: MenuItemProps) => (
  <div className={`menu-item ${selected ? ' selected' : ''}`} onClick={onClick}>
    <div className="menu-item-icon">{icon}</div>
    <span className="menu-item-title">{title}</span>
  </div>
);

MenuItem.defaultProps = {
  selected: false,
};

interface MenuProps {
  items: {
    title: string;
    icon: React.ReactNode;
  }[];
  onClick: (id: number) => void;
  selected: number;
}

export const Menu: React.FunctionComponent<MenuProps> = ({
  items,
  selected,
  onClick,
}: MenuProps) => (
  <section className="menu">
    {items.map((menuItem, index) => (
      <MenuItem
        key={index}
        {...menuItem}
        onClick={(): void => onClick(index)}
        selected={index === selected}
      />
    ))}
  </section>
);
