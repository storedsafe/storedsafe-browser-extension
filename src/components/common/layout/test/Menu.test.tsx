import * as React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import pretty from 'pretty'
import { Menu, MenuItem } from '../Menu'

/**
 * Set up and tear down container to mount tested component in.
 * */
let container: Element = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})
afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

const fooClick = jest.fn()
const barClick = jest.fn()

const items: MenuItem[] = [
  { title: 'foo', icon: <svg>icon</svg>, onClick: fooClick },
  { title: 'bar', icon: <p>icon</p>, onClick: barClick }
]

beforeEach(() => {
  fooClick.mockClear()
  barClick.mockClear()
})

test('<Menu show={false} />', () => {
  act(() => {
    render(<Menu items={items} show={false} />, container)
  })
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test('<Menu show={true} />', () => {
  act(() => {
    render(<Menu items={items} show={true} />, container)
  })
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test('<Menu />', () => {
  act(() => {
    render(<Menu items={items} show={true} />, container)
  })

  const [foo, bar] = document.querySelectorAll('.menu-item')

  fooClick.mockClear()
  barClick.mockClear()
  act(() => {
    foo.dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })
  expect(fooClick).toHaveBeenCalledTimes(1)
  expect(barClick).not.toHaveBeenCalled()

  fooClick.mockClear()
  barClick.mockClear()
  act(() => {
    bar.dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })
  expect(fooClick).not.toHaveBeenCalled()
  expect(barClick).toHaveBeenCalledTimes(1)
})
