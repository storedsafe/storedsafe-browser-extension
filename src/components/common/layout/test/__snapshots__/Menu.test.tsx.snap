// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu show={false} /> 1`] = `
"<section class=\\"menu\\">
  <div class=\\"menu-backdrop\\">
    <div class=\\"menu-box\\"><button type=\\"button\\" class=\\"menu-item\\" tabindex=\\"-1\\">
        <p>foo</p><svg>icon</svg>
      </button><button type=\\"button\\" class=\\"menu-item\\" tabindex=\\"-1\\">
        <p>bar</p>
        <p>icon</p>
      </button></div>
  </div>
</section>"
`;

exports[`<Menu show={true} /> 1`] = `
"<section class=\\"menu show\\">
  <div class=\\"menu-backdrop\\">
    <div class=\\"menu-box\\"><button type=\\"button\\" class=\\"menu-item\\" tabindex=\\"0\\">
        <p>foo</p><svg>icon</svg>
      </button><button type=\\"button\\" class=\\"menu-item\\" tabindex=\\"0\\">
        <p>bar</p>
        <p>icon</p>
      </button></div>
  </div>
</section>"
`;
